include/openenclave/internal/vector.h

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef _OE_INTERNAL_VECTOR_H
+#define _OE_INTERNAL_VECTOR_H
+
+#include <openenclave/bits/defs.h>
+#include <openenclave/bits/result.h>
+#include <openenclave/bits/types.h>
+
+OE_EXTERNC_BEGIN
+
+typedef struct _oe_vector
+{
+    void* data;
+    size_t size;
+} oe_vector_t;
+
+/**
+ * Pack a vector array into a flat representation.
+ *
+ * This function packs an array of vector structures onto a flat buffer.
+ * Consider the following definition.
+ *
+ *     ```
+ *     oe_vector_t vectors[] =
+ *     {
+ *         { "red", 4 },
+ *         { "green", 6 },
+ *         { "blue", 5 },
+ *     };
+ *     size_t vector_count = sizeof(vectors) / sizeof(vectors[0]);
+ *     ```
+ *
+ * This is packed with the following call.
+ *
+ *     ```
+ *     void* buf;
+ *     size_t buf_size;
+ *     oe_vector_pack(vectors, vector_count, &buf, &buf_size);
+ *     ```
+ *
+ * The buffer layout for this example is shown below.
+ *
+ *     ```
+ *     [OFFSET0][4]                vector[0]
+ *     [OFFSET1][6]                vector[1]
+ *     [OOFSET2][5]                vector[2]
+ *     [red\0green\0blue\0]
+ *     ```
+ *
+ * Or more compactly:
+ *
+ *     ```
+ *     [3][OFFSET0][4][OFFSET1][6][OFFSET2][5][red\0green\0blue\0]
```

Comment:
The host can intentionally not null terminate the strings and serialize the buffer as:
`[3][OFFSET0][4][OFFSET1][6][OFFSET2][5][reddgreennbluee]`

Reply:
This has been removed.